### DISCUSSION TITLE
Add means for specifying the ArrayList base Index?

### DISCUSSION THREAD

**Discussion by GCuser99** [Theadd/ArrayList#3] [Ideas]
I've tested your ArrayList in tB - very awesome and exactly what I was looking for - a collection-like object that allows for modifying the members and more...

I'm wondering with tB's parameterized class constructor "New()", whether it would be possible and make sense to allow user to optionally specify the array base index using something like:

```vba
Dim list As ArrayList
Set list = New ArrayList(baseIndex:=1)

For i = 1 to list.Count
     'do something with list(i)
Next i
```
Thanks!

**Comment by Theadd** [comment]
Hi @GCuser99,

> whether it would be possible and make sense to allow user to optionally specify the array base index

This is something that I would have never added =) but, since using 1-based indexes is known to be a common practice in VB6/VBA, it makes sense.

I've implemented it already, it can be specified in the constructor just like you suggested and also by using the `.BaseIndex` property, which also allows to change the base index on-the-fly on an existing ArrayList. The `BaseIndex` value in an ArrayList is also propagated to the sub ranges returned by `.GetRange`, which can be set to a different `BaseIndex` than the ArrayList they point to.

Also, calling `.ToArray` on an ArrayList with a BaseIndex other than zero also generates a Variant() array with the lower and upper bounds corresponding to the current value of BaseIndex.

https://github.com/Theadd/ArrayList/pull/4/commits/10ea2bc8295aabb2e9aa57bb87631eaef7fd7657

Many portions of code have been modified to apply this feature and I've not quite tested them all, unit tests would've been great here... So, if you encounter any unexpected behaviour please post an issue.

Thank you! =)

**Comment by GCuser99** [comment]
@Theadd - thanks very much! 

When does your updated twin package make it into tB's package database? In my tB project, I removed the ArrayListLib package in references and then reimported but the resulting ArrayListLib is version 1.5.139.0, not 1.5.141.0. Am I doing something wrong?

Mike

**Comment by Theadd** [comment]
Hi @GCuser99 , my bad, they're actually the same tho, previous version was 1.4.X.0, i just had the BUILD version number to be auto incremented by twinBASIC each time I hit the **Build** button and since I'm building it as an ActiveX DLL and TWINPACK, both for Win32 and Win64 versions in every release, that number differs a little bit. You can see those 4 builds being pushed together here: https://github.com/Theadd/ArrayList/commit/54350806a1c4c418215d0125c8d976ae0c274d58

I'll move the auto increment to the revision number so it won't differ from the version released on github. <sup><small><em>(Major.Minor.Build.Revision)</em></small></sup>
